include/sparrow/allocator.hpp

```diff
@@ -0,0 +1,247 @@
+            }
+
+            std::unique_ptr<interface> clone() const override
+            {
+                return std::make_unique<impl<A>>(m_alloc);
+            }
+
+            bool equal(const interface& rhs) const override
+            {
+                if (auto* p = dynamic_cast<const impl<A>*>(&rhs))
+                {
+                    return p->m_alloc == m_alloc;
+                }
+                return false;
+            }
+        };
+
+        using storage_type = std::variant
+        <
+            std::allocator<T>,
+            std::pmr::polymorphic_allocator<T>,
+            std::unique_ptr<interface>
+        >;
+
+        template <class A>
+        std::unique_ptr<interface> make_storage(A&& alloc) const
+        {
+            return std::make_unique<impl<std::decay_t<A>>>(std::forward<A>(alloc));
+        }
+
+        template <class A>
+        requires is_value_stored<A>
+        A&& make_storage(A&& alloc) const
+        {
+            return std::forward<A>(alloc);
+        }
+
+        storage_type copy_storage(const storage_type& rhs) const
+        {
+            return std::visit([](auto&& arg) -> storage_type
+            {
+                using A = std::decay_t<decltype(arg)>;
+                if constexpr (is_value_stored<A>)
+                    return { A(arg) };
+                else
+                    return { arg->clone() };
+            }, rhs);
+        }
+
+        storage_type m_storage;
+    };
+
+    /********************************
+     * any_allocator implementation *
+     ********************************/
+
+    template <class T, class DA>
+    any_allocator<T, DA>::any_allocator()
+        : m_storage(make_storage(DA()))
+    {
+    }
+
+    template <class T, class DA>
+    any_allocator<T, DA>::any_allocator(const any_allocator& rhs)
+        : m_storage(copy_storage(rhs.m_storage))
+    {
+    }
+
+    template <class T, class DA>
+    [[nodiscard]] T* any_allocator<T, DA>::allocate(std::size_t n)
+    {
+        return std::visit([n](auto&& arg)
+        {
+            using A = std::decay_t<decltype(arg)>;
+            if constexpr (is_value_stored<A>)
+                return arg.allocate(n);
+            else
+                return arg->allocate(n);
+        }, m_storage);
+    }
+
+    template <class T, class DA>
+    void any_allocator<T, DA>::deallocate(T* p, std::size_t n)
+    {
+        return std::visit([p, n](auto&& arg)
+        {
+            using A = std::decay_t<decltype(arg)>;
+            if constexpr (is_value_stored<A>)
+                arg.deallocate(p, n);
+            else
+                arg->deallocate(p, n);
+        }, m_storage);
+    }
+
+    template <class T, class DA>
+    any_allocator<T, DA> any_allocator<T, DA>::select_on_container_copy_construction() const
+    {
+        return any_allocator(*this);
+    }
+
+    template <class T, class DA>
+    bool any_allocator<T, DA>::equal(const any_allocator& rhs) const
```

Comment:
I'm not sure we even need allocator comparison? Usually it's only useful to be able to see what's the type of the allocator, it doesnt really have an observable state (at least standard allocators dont).

Reply:
ok that's part of standard requiremetns